File: stuff_data_gateway/src/main/java/android/montreal/devfest/stuff/data/gateway/DataGatewayModule.java
package android.montreal.devfest.stuff.data.gateway;

import android.montreal.devfest.stuff.usescases.RandomStuffGetter;
import android.montreal.devfest.stuff.usescases.RandomStuffSetter;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

@Module
public abstract class DataGatewayModule {

    @Provides
    @Singleton
    static DataGateway dataGateway(StuffStorage stuffStorage) {
        return new DataGateway(stuffStorage);
    }

    @Binds
    abstract RandomStuffGetter randomStuffGetter(DataGateway dataGateway);

    @Binds
    abstract RandomStuffSetter randomStuffSetter(DataGateway dataGateway);
}

File: stuff_data_gateway/src/main/java/android/montreal/devfest/stuff/data/gateway/DataGatewayModule.kt
package android.montreal.devfest.stuff.data.gateway

import android.montreal.devfest.stuff.usecases.RandomStuffGetter
import android.montreal.devfest.stuff.usecases.RandomStuffSetter
import dagger.Binds
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
abstract class DataGatewayModule {
    @Binds
    internal abstract fun randomStuffGetter(stuffDataGateway: StuffDataGateway): RandomStuffGetter

    @Binds
    internal abstract fun randomStuffSetter(stuffDataGateway: StuffDataGateway): RandomStuffSetter

    @Module
    companion object {
        @Provides
        @Singleton
        @JvmStatic
        internal fun dataGateway(stuffStorage: StuffStorage): StuffDataGateway {
            return StuffDataGateway(stuffStorage)
        }
    }
}
